File: app/src/main/java/pl/trikimusic/controller/core/gesture/FullRotationGestureDetector.kt
package pl.trikimusic.controller.core.gesture

import kotlin.math.abs
import kotlin.math.cos
import pl.trikimusic.controller.domain.model.FilteredSensorData
import pl.trikimusic.controller.domain.model.MediaAction
import pl.trikimusic.controller.domain.model.Vector3

/**
 * Recognizes one deliberate 270° rotation around the capsule's local Z axis while it is inverted.
 *
 * The detector intentionally follows the same signed gyro integration model as the volume
 * controller. A positive Z rotation is a right turn, a negative Z rotation is a left turn. The
 * inverted pose and acceleration magnitude gate keep track navigation separate from volume use.
 */
class FullRotationGestureDetector(
    private val configuration: Configuration = Configuration(),
) {
    data class Configuration(
        val stabilizationMillis: Long = 500L,
        // The two low-pass stages lose roughly 20–25° at the edges of a deliberate turn. A 245°
        // integrated threshold therefore makes the physical gesture finish around the requested 270°.
        val requiredRotationDegrees: Float = FILTERED_ROTATION_TRIGGER_DEGREES,
        val maximumRotationDegrees: Float = 420f,
        val maximumRotationMillis: Long = 5_000L,
        val maximumFaceDownTiltDegrees: Float = 25f,
        val maximumAccelerationDeviationG: Float = 0.20f,
        val activationGyroscopeDps: Float = 22f,
        val releaseGyroscopeDps: Float = 12f,
        val gyroscopeSmoothingAlpha: Float = 0.16f,
        val maximumSampleGapMillis: Long = 250L,
        val armingMaximumAngularRateDps: Float = 45f,
        val rearmQuietMillis: Long = 180L,
    ) {
        init {
            require(stabilizationMillis in 200L..3_000L)
            require(requiredRotationDegrees.isFinite() && requiredRotationDegrees in 180f..360f)
            require(maximumRotationDegrees.isFinite() && maximumRotationDegrees > requiredRotationDegrees)
            require(maximumRotationMillis in 1_000L..10_000L)
            require(maximumFaceDownTiltDegrees.isFinite() && maximumFaceDownTiltDegrees in 5f..45f)
            require(maximumAccelerationDeviationG.isFinite() && maximumAccelerationDeviationG in 0.05f..0.50f)
            require(activationGyroscopeDps.isFinite() && activationGyroscopeDps > 0f)
            require(releaseGyroscopeDps.isFinite() && releaseGyroscopeDps in 0f..activationGyroscopeDps)
            require(gyroscopeSmoothingAlpha.isFinite() && gyroscopeSmoothingAlpha in 0.01f..1f)
            require(maximumSampleGapMillis in 100L..1_000L)
            require(armingMaximumAngularRateDps.isFinite() && armingMaximumAngularRateDps in 10f..120f)
            require(rearmQuietMillis in 80L..500L)
        }
    }

    private var previousTimestampNanos: Long? = null
    private var stabilizationSinceNanos: Long? = null
    private var smoothedGyroscopeZ: Float? = null
    private var accumulatedRotationDegrees = 0f
    private var motionStartedNanos: Long? = null
    private var activeDirection = 0
    private var awaitingQuietRearm = false
    private var quietRearmSinceNanos: Long? = null
    private var triggered = false
    private var faceDown = false

    fun reset() {
        previousTimestampNanos = null
        stabilizationSinceNanos = null
        smoothedGyroscopeZ = null
        accumulatedRotationDegrees = 0f
        motionStartedNanos = null
        activeDirection = 0
        awaitingQuietRearm = false
        quietRearmSinceNanos = null
        triggered = false
        faceDown = false
    }

    fun process(sample: FilteredSensorData): FullRotationGestureResult {
        val timestampNanos = sample.source.timestampNanos
        val acceleration = sample.accelerometerG
        if (!isUsableAcceleration(acceleration)) {
            restartStabilization(timestampNanos, null)
            return result(HoldGesturePhase.HOLDING, 0f)
        }
        faceDown = isFaceDown(acceleration)

        val previousTimestamp = previousTimestampNanos
        if (
            previousTimestamp != null &&
            (timestampNanos <= previousTimestamp ||
                timestampNanos - previousTimestamp > configuration.maximumSampleGapMillis * NANOS_PER_MILLISECOND)
        ) {
            restartStabilization(timestampNanos, acceleration)
            return result(HoldGesturePhase.HOLDING, 0f)
        }
        previousTimestampNanos = timestampNanos

        val stabilizationStart = stabilizationSinceNanos
        if (stabilizationStart == null) {
            stabilizationSinceNanos = timestampNanos
            return result(HoldGesturePhase.HOLDING, 0f)
        }

        if (!faceDown || abs(sample.accelerationMagnitude - STANDARD_GRAVITY_G) > configuration.maximumAccelerationDeviationG) {
            restartStabilization(timestampNanos, acceleration)
            return result(HoldGesturePhase.HOLDING, 0f)
        }

        val stabilizationElapsedNanos = (timestampNanos - stabilizationStart).coerceAtLeast(0L)
        val requiredStabilizationNanos = configuration.stabilizationMillis * NANOS_PER_MILLISECOND
        val stabilizationProgress = if (requiredStabilizationNanos == 0L) {
            1f
        } else {
            (stabilizationElapsedNanos.toDouble() / requiredStabilizationNanos).toFloat().coerceIn(0f, 1f)
        }
        if (stabilizationElapsedNanos < requiredStabilizationNanos) {
            resetRotation()
            return result(HoldGesturePhase.HOLDING, stabilizationProgress)
        }

        val gyroscopeZ = smoothGyroscopeZ(sample.gyroscopeDps.z)
        if (awaitingQuietRearm) {
            if (
                abs(gyroscopeZ) <= configuration.releaseGyroscopeDps &&
                sample.gyroscopeMagnitude <= configuration.armingMaximumAngularRateDps
            ) {
                val quietSince = quietRearmSinceNanos ?: timestampNanos.also { quietRearmSinceNanos = it }
                if (timestampNanos - quietSince >= configuration.rearmQuietMillis * NANOS_PER_MILLISECOND) {
                    restartStabilization(timestampNanos, acceleration)
                }
            } else {
                quietRearmSinceNanos = null
            }
            return result(HoldGesturePhase.REARMING, 1f, gyroscopeZ)
        }

        if (!gyroscopeZ.isFinite()) {
            invalidateMotion(timestampNanos, acceleration)
            return result(HoldGesturePhase.REARMING, 1f)
        }
        if (abs(gyroscopeZ) <= configuration.releaseGyroscopeDps) {
            resetRotation(preserveSmoothing = true)
            return result(HoldGesturePhase.READY, 1f, gyroscopeZ)
        }

        val direction = if (gyroscopeZ > 0f) 1 else -1
        if (activeDirection != 0 && activeDirection != direction) {
            resetRotation(preserveSmoothing = true)
        }
        if (activeDirection == 0 && abs(gyroscopeZ) < configuration.activationGyroscopeDps) {
            return result(HoldGesturePhase.READY, 1f, gyroscopeZ)
        }
        activeDirection = direction
        if (motionStartedNanos == null) motionStartedNanos = timestampNanos
        val motionElapsedNanos = timestampNanos - (motionStartedNanos ?: timestampNanos)
        if (motionElapsedNanos > configuration.maximumRotationMillis * NANOS_PER_MILLISECOND) {
            invalidateMotion(timestampNanos, acceleration)
            return result(HoldGesturePhase.REARMING, 1f, gyroscopeZ)
        }

        val deltaSeconds = if (previousTimestamp == null || timestampNanos <= previousTimestamp) {
            0f
        } else {
            (timestampNanos - previousTimestamp).coerceAtMost(MAX_SAMPLE_INTERVAL_NANOS) / NANOS_PER_SECOND
        }
        val nextRotation = accumulatedRotationDegrees + gyroscopeZ * deltaSeconds
        if (abs(nextRotation) > configuration.maximumRotationDegrees) {
            invalidateMotion(timestampNanos, acceleration)
            return result(HoldGesturePhase.REARMING, 1f, gyroscopeZ)
        }
        accumulatedRotationDegrees = nextRotation

        val progress = (abs(accumulatedRotationDegrees) / configuration.requiredRotationDegrees).coerceIn(0f, 1f)
        if (abs(accumulatedRotationDegrees) >= configuration.requiredRotationDegrees) {
            triggered = true
            awaitingQuietRearm = true
            quietRearmSinceNanos = null
            return result(
                phase = HoldGesturePhase.TRIGGERED,
                stabilizationProgress = 1f,
                gyroscopeZ = gyroscopeZ,
                direction = if (accumulatedRotationDegrees > 0f) RotationGestureDirection.RIGHT else RotationGestureDirection.LEFT,
            )
        }

        return result(
            phase = HoldGesturePhase.TRACKING,
            stabilizationProgress = progress,
            gyroscopeZ = gyroscopeZ,
        )
    }

    private fun result(
        phase: HoldGesturePhase,
        stabilizationProgress: Float,
        gyroscopeZ: Float = smoothedGyroscopeZ ?: 0f,
        direction: RotationGestureDirection? = null,
    ) = FullRotationGestureResult(
        triggered = triggered && phase == HoldGesturePhase.TRIGGERED,
        direction = direction ?: activeDirection.takeIf { it != 0 }?.let {
            if (it > 0) RotationGestureDirection.RIGHT else RotationGestureDirection.LEFT
        },
        phase = phase,
        stabilizationProgress = stabilizationProgress,
        faceDown = faceDown,
        estimatedRotationDegrees = accumulatedRotationDegrees,
        gyroscopeZDps = gyroscopeZ.takeIf(Float::isFinite) ?: 0f,
    )

    private fun resetRotation(preserveSmoothing: Boolean = false) {
        accumulatedRotationDegrees = 0f
        motionStartedNanos = null
        activeDirection = 0
        if (!preserveSmoothing) smoothedGyroscopeZ = null
    }

    private fun restartStabilization(timestampNanos: Long, acceleration: Vector3?) {
        previousTimestampNanos = timestampNanos
        stabilizationSinceNanos = timestampNanos
        faceDown = acceleration?.let(::isFaceDown) ?: false
        awaitingQuietRearm = false
        quietRearmSinceNanos = null
        triggered = false
        resetRotation()
    }

    private fun invalidateMotion(timestampNanos: Long, acceleration: Vector3) {
        previousTimestampNanos = timestampNanos
        stabilizationSinceNanos = timestampNanos
        faceDown = isFaceDown(acceleration)
        awaitingQuietRearm = true
        quietRearmSinceNanos = null
        triggered = false
        resetRotation(preserveSmoothing = true)
    }

    private fun smoothGyroscopeZ(value: Float): Float {
        val previous = smoothedGyroscopeZ
        val smoothed = if (previous == null) value else {
            previous + configuration.gyroscopeSmoothingAlpha * (value - previous)
        }
        smoothedGyroscopeZ = smoothed
        return smoothed
    }

    private fun isFaceDown(value: Vector3): Boolean {
        val magnitude = value.magnitude
        if (!magnitude.isFinite() || magnitude < MIN_USABLE_ACCELERATION_G) return false
        val zComponent = (value.z / magnitude).coerceIn(-1f, 1f)
        return zComponent >= cos(Math.toRadians(configuration.maximumFaceDownTiltDegrees.toDouble())).toFloat()
    }

    private fun isUsableAcceleration(value: Vector3): Boolean =
        value.x.isFinite() && value.y.isFinite() && value.z.isFinite() &&
            value.magnitude in MIN_USABLE_ACCELERATION_G..MAX_USABLE_ACCELERATION_G

    companion object {
        const val PHYSICAL_ROTATION_TARGET_DEGREES = 270f
        const val FILTERED_ROTATION_TRIGGER_DEGREES = 245f

        private const val NANOS_PER_MILLISECOND = 1_000_000L
        private const val NANOS_PER_SECOND = 1_000_000_000f
        private const val MAX_SAMPLE_INTERVAL_NANOS = 100_000_000L
        private const val STANDARD_GRAVITY_G = 1f
        private const val MIN_USABLE_ACCELERATION_G = 0.20f
        private const val MAX_USABLE_ACCELERATION_G = 2.50f
    }
}

data class FullRotationGestureResult(
    val triggered: Boolean = false,
    val direction: RotationGestureDirection? = null,
    val phase: HoldGesturePhase,
    val stabilizationProgress: Float,
    val faceDown: Boolean,
    val estimatedRotationDegrees: Float,
    val gyroscopeZDps: Float,
)

enum class RotationGestureDirection {
    RIGHT,
    LEFT,
}

fun RotationGestureDirection.toInvertedCapsuleNavigationAction(): MediaAction = when (this) {
    RotationGestureDirection.LEFT -> MediaAction.NEXT
    RotationGestureDirection.RIGHT -> MediaAction.PREVIOUS
}

File: windows/TrikiMusicController.Windows/Core/FullRotationGestureDetector.cs
using TrikiMusicController_Windows.Models;

namespace TrikiMusicController_Windows.Core;

/// <summary>
/// Recognizes one deliberate 270° rotation around the inverted capsule's local Z axis.
/// </summary>
public sealed record FullRotationGestureConfiguration(
    long StabilizationMillis = 500,
    // The two low-pass stages lose roughly 20–25° at the edges of a deliberate turn. A 245°
    // integrated threshold therefore makes the physical gesture finish around the requested 270°.
    float RequiredRotationDegrees = FullRotationGestureDetector.FilteredRotationTriggerDegrees,
    float MaximumRotationDegrees = 420,
    long MaximumRotationMillis = 5_000,
    float MaximumFaceDownTiltDegrees = 25,
    float MaximumAccelerationDeviationG = 0.20f,
    float ActivationGyroscopeDps = 22,
    float ReleaseGyroscopeDps = 12,
    float GyroscopeSmoothingAlpha = 0.16f,
    long MaximumSampleGapMillis = 250,
    float ArmingMaximumAngularRateDps = 45,
    long RearmQuietMillis = 180);

public sealed class FullRotationGestureDetector
{
    public const float PhysicalRotationTargetDegrees = 270;
    public const float FilteredRotationTriggerDegrees = 245;

    private const long NanosPerMillisecond = 1_000_000;
    private const long MaxSampleIntervalNanos = 100_000_000;
    private const float NanosPerSecond = 1_000_000_000f;
    private const float MinimumUsableAccelerationG = 0.20f;
    private const float MaximumUsableAccelerationG = 2.50f;
    private readonly FullRotationGestureConfiguration _configuration;
    private long? _previousTimestampNanos;
    private long? _stabilizationSinceNanos;
    private float? _smoothedGyroscopeZ;
    private float _accumulatedRotationDegrees;
    private long? _motionStartedNanos;
    private int _activeDirection;
    private bool _awaitingQuietRearm;
    private long? _quietRearmSinceNanos;
    private bool _triggered;
    private bool _faceDown;

    public FullRotationGestureDetector(FullRotationGestureConfiguration? configuration = null)
    {
        _configuration = configuration ?? new FullRotationGestureConfiguration();
        ValidateConfiguration(_configuration);
    }

    public void Reset()
    {
        _previousTimestampNanos = null;
        _stabilizationSinceNanos = null;
        _smoothedGyroscopeZ = null;
        _accumulatedRotationDegrees = 0;
        _motionStartedNanos = null;
        _activeDirection = 0;
        _awaitingQuietRearm = false;
        _quietRearmSinceNanos = null;
        _triggered = false;
        _faceDown = false;
    }

    public FullRotationGestureResult Process(FilteredSensorData sample)
    {
        var timestamp = sample.Source.TimestampNanos;
        var acceleration = sample.AccelerometerG;
        if (!IsUsableAcceleration(acceleration))
        {
            RestartStabilization(timestamp, null);
            return Result(HoldGesturePhase.Holding, 0);
        }
        _faceDown = IsFaceDown(acceleration);

        var previousTimestamp = _previousTimestampNanos;
        if (previousTimestamp is long previous &&
            (timestamp <= previous || timestamp - previous > _configuration.MaximumSampleGapMillis * NanosPerMillisecond))
        {
            RestartStabilization(timestamp, acceleration);
            return Result(HoldGesturePhase.Holding, 0);
        }
        _previousTimestampNanos = timestamp;

        var stabilizationStart = _stabilizationSinceNanos;
        if (stabilizationStart is null)
        {
            _stabilizationSinceNanos = timestamp;
            return Result(HoldGesturePhase.Holding, 0);
        }

        if (!_faceDown || Math.Abs(sample.AccelerationMagnitude - 1f) > _configuration.MaximumAccelerationDeviationG)
        {
            RestartStabilization(timestamp, acceleration);
            return Result(HoldGesturePhase.Holding, 0);
        }

        var stabilizationElapsed = Math.Max(0, timestamp - stabilizationStart.Value);
        var requiredStabilization = _configuration.StabilizationMillis * NanosPerMillisecond;
        var stabilizationProgress = requiredStabilization == 0
            ? 1f
            : Math.Clamp((float)((double)stabilizationElapsed / requiredStabilization), 0, 1);
        if (stabilizationElapsed < requiredStabilization)
        {
            ResetRotation();
            return Result(HoldGesturePhase.Holding, stabilizationProgress);
        }

        var gyroscopeZ = SmoothGyroscopeZ(sample.GyroscopeDps.Z);
        if (_awaitingQuietRearm)
        {
            if (Math.Abs(gyroscopeZ) <= _configuration.ReleaseGyroscopeDps &&
                sample.GyroscopeMagnitude <= _configuration.ArmingMaximumAngularRateDps)
            {
                _quietRearmSinceNanos ??= timestamp;
                if (timestamp - _quietRearmSinceNanos.Value >= _configuration.RearmQuietMillis * NanosPerMillisecond)
                    RestartStabilization(timestamp, acceleration);
            }
            else
            {
                _quietRearmSinceNanos = null;
            }
            return Result(HoldGesturePhase.Rearming, 1, gyroscopeZ);
        }

        if (!float.IsFinite(gyroscopeZ))
        {
            InvalidateMotion(timestamp, acceleration);
            return Result(HoldGesturePhase.Rearming, 1);
        }
        if (Math.Abs(gyroscopeZ) <= _configuration.ReleaseGyroscopeDps)
        {
            ResetRotation(preserveSmoothing: true);
            return Result(HoldGesturePhase.Ready, 1, gyroscopeZ);
        }

        var direction = gyroscopeZ > 0 ? 1 : -1;
        if (_activeDirection != 0 && _activeDirection != direction)
            ResetRotation(preserveSmoothing: true);
        if (_activeDirection == 0 && Math.Abs(gyroscopeZ) < _configuration.ActivationGyroscopeDps)
            return Result(HoldGesturePhase.Ready, 1, gyroscopeZ);

        _activeDirection = direction;
        _motionStartedNanos ??= timestamp;
        var motionElapsed = timestamp - _motionStartedNanos.Value;
        if (motionElapsed > _configuration.MaximumRotationMillis * NanosPerMillisecond)
        {
            InvalidateMotion(timestamp, acceleration);
            return Result(HoldGesturePhase.Rearming, 1, gyroscopeZ);
        }

        var deltaSeconds = previousTimestamp is null || timestamp <= previousTimestamp.Value
            ? 0
            : Math.Min(timestamp - previousTimestamp.Value, MaxSampleIntervalNanos) / NanosPerSecond;
        var nextRotation = _accumulatedRotationDegrees + gyroscopeZ * deltaSeconds;
        if (Math.Abs(nextRotation) > _configuration.MaximumRotationDegrees)
        {
            InvalidateMotion(timestamp, acceleration);
            return Result(HoldGesturePhase.Rearming, 1, gyroscopeZ);
        }
        _accumulatedRotationDegrees = nextRotation;

        var progress = Math.Clamp(Math.Abs(_accumulatedRotationDegrees) / _configuration.RequiredRotationDegrees, 0, 1);
        if (Math.Abs(_accumulatedRotationDegrees) >= _configuration.RequiredRotationDegrees)
        {
            _triggered = true;
            _awaitingQuietRearm = true;
            _quietRearmSinceNanos = null;
            return Result(
                HoldGesturePhase.Triggered,
                1,
                gyroscopeZ,
                _accumulatedRotationDegrees > 0 ? RotationGestureDirection.Right : RotationGestureDirection.Left);
        }

        return Result(HoldGesturePhase.Tracking, progress, gyroscopeZ);
    }

    private FullRotationGestureResult Result(
        HoldGesturePhase phase,
        float stabilizationProgress,
        float gyroscopeZ = 0,
        RotationGestureDirection? direction = null) => new(
        _triggered && phase == HoldGesturePhase.Triggered,
        direction ?? (_activeDirection == 0
            ? null
            : _activeDirection > 0 ? RotationGestureDirection.Right : RotationGestureDirection.Left),
        phase,
        stabilizationProgress,
        _faceDown,
        _accumulatedRotationDegrees,
        float.IsFinite(gyroscopeZ) ? gyroscopeZ : 0);

    private void ResetRotation(bool preserveSmoothing = false)
    {
        _accumulatedRotationDegrees = 0;
        _motionStartedNanos = null;
        _activeDirection = 0;
        if (!preserveSmoothing) _smoothedGyroscopeZ = null;
    }

    private void RestartStabilization(long timestamp, Vector3f? acceleration)
    {
        _previousTimestampNanos = timestamp;
        _stabilizationSinceNanos = timestamp;
        _faceDown = acceleration is Vector3f value && IsFaceDown(value);
        _awaitingQuietRearm = false;
        _quietRearmSinceNanos = null;
        _triggered = false;
        ResetRotation();
    }

    private void InvalidateMotion(long timestamp, Vector3f acceleration)
    {
        _previousTimestampNanos = timestamp;
        _stabilizationSinceNanos = timestamp;
        _faceDown = IsFaceDown(acceleration);
        _awaitingQuietRearm = true;
        _quietRearmSinceNanos = null;
        _triggered = false;
        ResetRotation(preserveSmoothing: true);
    }

    private float SmoothGyroscopeZ(float value)
    {
        var smoothed = _smoothedGyroscopeZ is float previous
            ? previous + _configuration.GyroscopeSmoothingAlpha * (value - previous)
            : value;
        _smoothedGyroscopeZ = smoothed;
        return smoothed;
    }

    private bool IsFaceDown(Vector3f value)
    {
        var magnitude = value.Magnitude;
        if (!float.IsFinite(magnitude) || magnitude < MinimumUsableAccelerationG) return false;
        var zComponent = Math.Clamp(value.Z / magnitude, -1, 1);
        return zComponent >= MathF.Cos(_configuration.MaximumFaceDownTiltDegrees * MathF.PI / 180f);
    }

    private static bool IsUsableAcceleration(Vector3f value) =>
        float.IsFinite(value.X) && float.IsFinite(value.Y) && float.IsFinite(value.Z) &&
        value.Magnitude is >= MinimumUsableAccelerationG and <= MaximumUsableAccelerationG;

    private static void ValidateConfiguration(FullRotationGestureConfiguration value)
    {
        if (value.StabilizationMillis is < 200 or > 3_000 ||
            !float.IsFinite(value.RequiredRotationDegrees) || value.RequiredRotationDegrees is < 180 or > 360 ||
            !float.IsFinite(value.MaximumRotationDegrees) || value.MaximumRotationDegrees <= value.RequiredRotationDegrees ||
            value.MaximumRotationMillis is < 1_000 or > 10_000 ||
            !float.IsFinite(value.MaximumFaceDownTiltDegrees) || value.MaximumFaceDownTiltDegrees is < 5 or > 45 ||
            !float.IsFinite(value.MaximumAccelerationDeviationG) || value.MaximumAccelerationDeviationG is < 0.05f or > 0.50f ||
            !float.IsFinite(value.ActivationGyroscopeDps) || value.ActivationGyroscopeDps <= 0 ||
            !float.IsFinite(value.ReleaseGyroscopeDps) || value.ReleaseGyroscopeDps < 0 || value.ReleaseGyroscopeDps > value.ActivationGyroscopeDps ||
            !float.IsFinite(value.GyroscopeSmoothingAlpha) || value.GyroscopeSmoothingAlpha is < 0.01f or > 1 ||
            value.MaximumSampleGapMillis is < 100 or > 1_000 ||
            !float.IsFinite(value.ArmingMaximumAngularRateDps) || value.ArmingMaximumAngularRateDps is < 10 or > 120 ||
            value.RearmQuietMillis is < 80 or > 500)
            throw new ArgumentOutOfRangeException(nameof(value));
    }
}
